src/backend/common/Framebuffer.cpp

```diff
@@ -0,0 +1,125 @@
+// Copyright 2017 The NXT Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "Framebuffer.h"
+
+#include "Buffer.h"
+#include "Device.h"
+#include "RenderPass.h"
+#include "Texture.h"
+
+namespace backend {
+
+    // Framebuffer
+
+    FramebufferBase::FramebufferBase(FramebufferBuilder* builder)
+        : renderPass(std::move(builder->renderPass)), width(builder->width), height(builder->height), textureViews(std::move(builder->textureViews)) {
+    }
+
+    RenderPassBase* FramebufferBase::GetRenderPass() {
+        return renderPass.Get();
+    }
+
+    TextureViewBase* FramebufferBase::GetTextureView(uint32_t index) {
+        ASSERT(index < textureViews.size());
+        return textureViews[index].Get();
+    }
+
+    // FramebufferBuilder
+
+    enum FramebufferSetProperties {
+        FRAMEBUFFER_PROPERTY_RENDER_PASS = 0x1,
+        FRAMEBUFFER_PROPERTY_DIMENSIONS = 0x2,
+    };
+
+    FramebufferBuilder::FramebufferBuilder(DeviceBase* device)
+        : Builder(device) {
+    }
+
+    FramebufferBase* FramebufferBuilder::GetResultImpl() {
+        constexpr int requiredProperties = FRAMEBUFFER_PROPERTY_RENDER_PASS | FRAMEBUFFER_PROPERTY_DIMENSIONS;
+        if ((propertiesSet & requiredProperties) != requiredProperties) {
+            HandleError("Framebuffer missing properties");
+            return nullptr;
+        }
+
+        for (uint32_t subpass = 0; subpass < renderPass->GetSubpassCount(); ++subpass) {
```

Comment:
Since the renderpass's attachments must all be defined, we only need to iterate over this->textureViews.